### ISSUE TITLE
Map structure

### ISSUE THREAD

**Issue by js-choi** [js-choi/proposal-function-memo#4]
If we go with a Map-like cache (see [policy Maps](https://github.com/js-choi/proposal-policy-map-set)), how should we structure the cache? For example, we could use a tree of Maps, or we could use argument-[tuples](https://github.com/tc39/proposal-record-tuple/blob/main/details.md) as keys in one Map.

There is also the [compositeKeys proposal](https://github.com/tc39/proposal-richer-keys/tree/master/compositeKey) (#1).

**Comment by michaelficarra** [comment]
Is it observable?

**Comment by js-choi** [comment]
@michaelficarra: It is observable if we expose the cache to the developer somehow. For example, we may allow the developer to supply a map-like cache argument. (See also [proposal-policy-map-set](https://github.com/js-choi/proposal-policy-map-set), which I will also present to plenary.)

**Comment by michaelficarra** [comment]
I don't like using a user-provided map. If we want memoised functions to have methods like `clear`, we can give them a prototype that extends `Function.prototype`.

**Comment by js-choi** [comment]
@michaelficarra: The more personally important use case for allowing custom caches is to allow the user to specify the memoization’s [cache replacement policy](https://en.wikipedia.org/wiki/Cache_replacement_policies). For example, I may want to specify that the cache store up to 100 entries with a least recently used (LRU) policy, like [how Python does](https://docs.python.org/3/library/functools.html#functools.lru_cache).

But, yes, hiding the cache’s structure completely is also an option for this issue.

**Comment by js-choi** [comment]
https://github.com/js-choi/proposal-function-memo/issues/5#issuecomment-1083466911 reminded me that records/tuples cannot directly contain objects, so that goes out the window. Looks like we would have to use trees of Map-likes after all.

**Comment by michaelficarra** [comment]
Once you choose a tree of map-likes, now you are giving the consumer per-argument-position control over the caching strategy, which I don't think anyone wants.

**Comment by zloirock** [comment]
Here you are trying to solve a problem that `compositeKeys` was intended to solve directly. My polyfill of `compositeKeys` represents the such tree of `WeakMap`s and `Map`s.

**Comment by js-choi** [comment]
@michaelficarra: I agree. I would much prefer a flat one-level cache data structure with complex keys. 

I suppose we could still:

1. Use argument tuples as keys in the developer-supplied map-like `cache`.
2. Replace objects in the tuples with symbols before calling `cache.has`, `cache.get`, or `cache.set`.
3. Store those symbols as keys in a separate symbol→object WeakMap, which would not be exposed to the developer.

For example:

```js
const cache = new Map;
const f = (function f (arg0) { return this.toString() + arg0.toString(); }).memo(cache);
const o0 = { x: 0 }, o1 = { x: 1 };
f.call(o0, 0); // Returns '{ x: 0 }0'.
f.call(o1, 1); // Returns '{ x: 1 }1'.
```

Now `cache` would be `Map(2) { #[s0, undefined, 0] ⇒ '{ x: 0 }0', #[s1, undefined, 1] ⇒ '{ x: 1 }1' }`, where `s0` and `s1` are unique symbols, and `f` contains a WeakMap `{ s0 ⇒ o0, s1 ⇒ o1 }`.

This would depend on [symbols as WeakMap keys](https://github.com/tc39/proposal-symbols-as-weakmap-keys), which is hopefully supposed to reach Stage 3 soon. This is one of its core use cases, after all.

***

@zloirock: Although `compositeKey` would be great for this, it has been stuck at Stage 1 with no progress for two years, so I am concerned that it will be a long time before it gets added to the language (if ever).

At plenary, I will present compositeKey as an alternative possibility to using tuple keys containing symbols in a WeakMap. We will discuss both approaches.

**Comment by zloirock** [comment]
> Although compositeKey would be great for this, it has been stuck at Stage 1 with no progress for two years, so I am concerned that it will be a long time before it gets added to the language (if ever).

This proposal is a good chance to advance `compositeKey`.

**Comment by js-choi** [comment]
According to [composite keys’ explainer](https://github.com/tc39/proposal-richer-keys/tree/master/compositeKey), “at least one component must be a valid key that can be placed in a WeakMap”. Unless I am mistaken, this restriction would exclude them from being usable for argument-list keys, since both `compositeKey()` (for `f()`) and (`compositeKey(1)` for `f(1)`) would be invalid. 

We still have the “replace objects with symbols from a WeakMap” solution, though.

**Comment by zloirock** [comment]
@js-choi a first key, for example, can be a collection.

**Comment by theScottyJam** [comment]
A Map tree would both be difficult to manage if you ever want to touch it by hand in userland, and may prevent a built-in LRU-Map from working, unless it's also built to support this tree shape.

I think an ideal default option would be to use a tuple for the cache key (update: I discuss more on how I think this is still possible despite previous concerns near the bottom). I also feel that this should be completely customizable, as I don't think it's possible to provide an algorithm that would satisfy everyone, and requiring people to customize the algorithm on a custom Map instance itself would be tedious, and make many use cases of function memoization overly difficult.

An option I've used previously is to require the user to generate a "cache key" from the provided parameters. So, something like this:

```javascript
const add = (x, y) => x + y
const memoizedAdd = memoize(add, {
  argsToKey: (...args) => JSON.stringify(args)
})
```

In this example, if you call `memoizedAdd` with `2` and `3`, the engine will call `argsToKey` with those parameters to determine what should be used as the map key. In this case, the key will be `'[2,3]'`.

This puts a ton of power into the hands of the user - power which is often needed. It can solve scenarios such as these:
* What if the function expects an options bag as an argument? There's no resonable defaults you can do to peer into objects and compare their values (there's too many issues related to prototypes, circular references, private fields, etc). An `argsToKey` function allows the user to hand-pluck important properties from an object bag and generate a key from it.
* What if you want to treat two types of values as equivalent? e.g. perahps the second parameter of your function will always behave the same if it's set to `null`, `undefined`, it's missing, or you provide the value it would have used as the default anyways. Or, you can imagine a function that takes an angle's degrees as one of its parameters, and perhaps you want to normalize the numbers so they're in the range 0-360 before you use them as a cache key (e.g. if 361 is passed in, you don't use 361 as a cache key, you just use 1).
* You can ignore arguments that don't need to be part of the cache key. For example, imagine memoizing a function that takes an Express request instance as a parameter. The function being memoized will pluck some query parameters and headers off of the request instance, but that's it, it doesn't care what endpoint was hit to trigger this request or anything else that could also be found on this instance. With a custom argsToKey function, you have the power to pluck what you want off of the req object and incorporate it as part of the cache key, then ignore everything else that's irrelevant to the function being memoized.
* We can use the `Tuple` constructor as the default for the argsToKey function once it's available (I think that would make the most sense as a default), however, we don't have to tie this proposal to the Tuple proposal. To start with, we could make it an error to use the memoize function without providing an argsToKey function. A follow-on proposal can change the default to something else when we're ready for it. (Update: Sorry, only quickly skimmed over this conversation, but reading through it again, I noticed there was concern about the fact that objects can't go into tuples. We could choose to auto-box objects so they can go in them, or, we could choose to throw an error if an object is provided - is the user doing the memoization wants to support objects, they have to supply their own argsToKey function).

**Comment by js-choi** [comment]
It is a good point that key customization is important, and that the tuple-based approach in https://github.com/tc39/proposal-function-memo/issues/4#issuecomment-1083552333 would not address even common use cases like option objects.

It is also a good point that we can punt this problem to the future by making keying functions be required. It’s not a big deal to write something like `fn.memo(memoCache, ([ key ]) => key)`. This would decouple this proposal from tuples (or compositeKeys). We could always add a default keying function in the future.
